doc/design/error_reporting_api.md

```diff
@@ -0,0 +1,240 @@
+# The New Error Reporting API
+
+This is intended as a quick guide through the new error reporting API that will gradually replace
+the previous error handler.
+
+## Terminology
+
+### Module 
+
+Self-contained part of code, e.g. hoofs or posh (or smaller scope). Each module must have a unique
+id and defines its errors.
+
+### Module Id
+
+A unique id for each module that is defined by some central instance that knows all modules that are
+to be used in some project.
+
+### Platform 
+
+The backend code that is used by any module. It can be changed at compile time and there is only
+one active platform at a time.
+
+### Error Reporting API
+
+The frontend to repor errors. API for short.
+
+### Error Handler
+
+The backend that defines the actions on reported errors. Defined by the platform.
+
+### Error Code
+
+Some error code that is unique per module.
+
+### Error
+
+An error object to be reported. Must contain at least some error code and is associated with some
+module (by its id).
+
+## How it works
+
+The whole mechanism is very flexible and generic and relies on overloading and perfect forwarding 
+to relay information, e.g. errors. That means much of it can be optimized at compile time, 
+as the compiler has full knowledge about all the template definitions. 
+
+This holds up to the platform, where it is possible to e.g. use runtime polymorphism or simply link
+against some library.
+
+## Reponsibilty of the Platform
+
+1. Define the error handling backend
+1. It expects that errors satisfy some mild conditions (like providing a code)
+1. Provide a header that defines the reporting; to be used by any module
+
+In principle a platform can support a wide range of error types, from simple code to monadic result
+types. The only condition is that the backend must be able to handle them.
+
+In this way, the platform defines a contract for the errors.
+
+## Reponsibilty of a Module
+
+1. Define all its errors and error codes
+1. The errors must satisfy the error contract conditions required by the platform.
+1. Define how codes are converted to errors
+1. Combine the definitions with the API and rovide a header that defines its whole error reporting.
+
+Once this is defined, the module can use the error reporting in its own functions.
+
+## Kinds of Errors
+
+There are three mandatory kinds (or categories) of errors.
+
+1. Fatal errors
+1. Precondition violations
+1. Debug assert violations
+
+All of them must be supported by the platform and all of them abort execution (which cannot be
+changed).
+
+Furthermore the platform may define additional kinds of errors, specifically non-fatal errors.
+For non-fatal errors the execution continues afer reporting, so they should be used when the
+circumstances allow recovery from the error.
```

Comment:
This is intended for what we previously had in posh as warnings etc.

My idea is to support three steps.

1. Generate the error `e` somehow (`IOX_ERROR`, code, get it by return value from somewhere etc.)
2. Report, but maybe not fatal.
3. If not fatal, control flow continues (likely differently) and `e` is returned or thrown or whatever.

Maybe there is a misunderstanding. The platform can add custom kinds, the user is not supposed to do this. They have to use what the platform provides, otherwise the code will not compile. This is similar to trying to throw a `std::runtime_error` but it is not defined. We rely on this being supported by the "platform", the C++ lib in this case.

Reply:
It is not only the user which has to use whatever the platform provides but every platform hast to provide whatever the user uses. Not specifying this repeats one of the big mistakes of C++ by leaving stuff unspecified and let the vendors decide. Neither the user nor the provider of the platform has full freedom and by not specifying it we make their life harder